File: infrastructure/go-sdk/redis.go
package infra

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/redis/go-redis/v9"
	"go.uber.org/zap"
)

type RedisClient struct {
	client *redis.Client
	logger *zap.Logger
}

func NewRedisClient(logger *zap.Logger, addr string) *RedisClient {
	c := &RedisClient{logger: logger}
	rdb := redis.NewClient(&redis.Options{
		Addr:            addr,
		PoolSize:        20,
		MinIdleConns:    5,
		MaxRetries:      3,
		DialTimeout:     3 * time.Second,
		ReadTimeout:     2 * time.Second,
		WriteTimeout:    2 * time.Second,
		PoolTimeout:     3 * time.Second,
		ConnMaxLifetime: 30 * time.Minute,
	})
	c.client = rdb

	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
	defer cancel()
	if err := rdb.Ping(ctx).Err(); err != nil {
		logger.Warn("redis_ping_failed", zap.Error(err))
	} else {
		logger.Info("redis_connected", zap.String("addr", addr))
	}
	return c
}

func (c *RedisClient) Ping(ctx context.Context) error {
	if c.client == nil {
		return fmt.Errorf("redis not initialized")
	}
	return c.client.Ping(ctx).Err()
}

func (c *RedisClient) Client() *redis.Client { return c.client }

func (c *RedisClient) CacheJSON(ctx context.Context, key string, value interface{}, ttl time.Duration) error {
	data, err := json.Marshal(value)
	if err != nil {
		return fmt.Errorf("marshal: %w", err)
	}
	return c.client.Set(ctx, key, data, ttl).Err()
}

func (c *RedisClient) GetCachedJSON(ctx context.Context, key string, dest interface{}) error {
	data, err := c.client.Get(ctx, key).Bytes()
	if err != nil {
		return err
	}
	return json.Unmarshal(data, dest)
}

func (c *RedisClient) RateLimit(ctx context.Context, key string, maxRequests int64, window time.Duration) (bool, error) {
	pipe := c.client.Pipeline()
	incr := pipe.Incr(ctx, key)
	pipe.Expire(ctx, key, window)
	_, err := pipe.Exec(ctx)
	if err != nil {
		return false, err
	}
	return incr.Val() <= maxRequests, nil
}

func (c *RedisClient) AcquireLock(ctx context.Context, key string, ttl time.Duration) (bool, error) {
	return c.client.SetNX(ctx, "lock:"+key, time.Now().UnixMilli(), ttl).Result()
}

func (c *RedisClient) ReleaseLock(ctx context.Context, key string) error {
	return c.client.Del(ctx, "lock:"+key).Err()
}

func (c *RedisClient) Publish(ctx context.Context, channel string, message interface{}) error {
	data, err := json.Marshal(message)
	if err != nil {
		return err
	}
	return c.client.Publish(ctx, channel, data).Err()
}

func (c *RedisClient) Subscribe(ctx context.Context, channel string) *redis.PubSub {
	return c.client.Subscribe(ctx, channel)
}

func (c *RedisClient) SetKYCGate(ctx context.Context, userID string, allowed bool, level int, ttl time.Duration) error {
	data := map[string]interface{}{"allowed": allowed, "level": level, "ts": time.Now().Unix()}
	return c.CacheJSON(ctx, "kyc:gate:"+userID, data, ttl)
}

func (c *RedisClient) GetKYCGate(ctx context.Context, userID string) (bool, int, error) {
	var data map[string]interface{}
	if err := c.GetCachedJSON(ctx, "kyc:gate:"+userID, &data); err != nil {
		return false, 0, err
	}
	allowed, _ := data["allowed"].(bool)
	level := int(data["level"].(float64))
	return allowed, level, nil
}

func (c *RedisClient) InvalidatePattern(ctx context.Context, pattern string) (int64, error) {
	var cursor uint64
	var deleted int64
	for {
		keys, nextCursor, err := c.client.Scan(ctx, cursor, pattern, 100).Result()
		if err != nil {
			return deleted, err
		}
		if len(keys) > 0 {
			d, _ := c.client.Del(ctx, keys...).Result()
			deleted += d
		}
		cursor = nextCursor
		if cursor == 0 {
			break
		}
	}
	return deleted, nil
}

func (c *RedisClient) PoolStats() *redis.PoolStats {
	if c.client == nil {
		return nil
	}
	return c.client.PoolStats()
}

func (c *RedisClient) Close() {
	if c.client != nil {
		c.client.Close()
	}
}

File: infrastructure/ts-sdk/src/redis.ts
/**
 * Redis client with connection pooling, rate limiting, KYC gate, and pub/sub.
 */

export class RedisClient {
  private addr: string;
  private client: any;

  constructor(addr: string) {
    this.addr = addr;
    try {
      const Redis = require('ioredis');
      const [host, port] = addr.split(':');
      this.client = new Redis({ host, port: parseInt(port || '6379'), maxRetriesPerRequest: 3, retryStrategy: (times: number) => Math.min(times * 100, 3000) });
    } catch { this.client = null; }
  }

  async ping(): Promise<void> {
    if (!this.client) throw new Error('Redis not initialized');
    await this.client.ping();
  }

  async cacheJSON(key: string, value: unknown, ttlSeconds: number = 300): Promise<void> {
    if (!this.client) return;
    await this.client.set(key, JSON.stringify(value), 'EX', ttlSeconds);
  }

  async getCachedJSON<T = unknown>(key: string): Promise<T | null> {
    if (!this.client) return null;
    const data = await this.client.get(key);
    return data ? JSON.parse(data) : null;
  }

  async rateLimit(key: string, maxRequests: number, windowSeconds: number): Promise<boolean> {
    if (!this.client) return true;
    const pipeline = this.client.pipeline();
    pipeline.incr(key);
    pipeline.expire(key, windowSeconds);
    const results = await pipeline.exec();
    return (results?.[0]?.[1] ?? 0) <= maxRequests;
  }

  async acquireLock(key: string, ttlSeconds: number = 30): Promise<boolean> {
    if (!this.client) return false;
    const result = await this.client.set(`lock:${key}`, Date.now(), 'NX', 'EX', ttlSeconds);
    return result === 'OK';
  }

  async releaseLock(key: string): Promise<void> {
    if (!this.client) return;
    await this.client.del(`lock:${key}`);
  }

  async publish(channel: string, message: unknown): Promise<void> {
    if (!this.client) return;
    await this.client.publish(channel, JSON.stringify(message));
  }

  async setKYCGate(userId: string, allowed: boolean, level: number, ttl: number = 600): Promise<void> {
    await this.cacheJSON(`kyc:gate:${userId}`, { allowed, level, ts: Math.floor(Date.now() / 1000) }, ttl);
  }

  async getKYCGate(userId: string): Promise<{ allowed: boolean; level: number } | null> {
    return this.getCachedJSON(`kyc:gate:${userId}`);
  }

  async cachePolicy(policyId: string, data: Record<string, unknown>, ttl: number = 3600): Promise<void> {
    await this.cacheJSON(`policy:${policyId}`, data, ttl);
  }

  async getCachedPolicy(policyId: string): Promise<Record<string, unknown> | null> {
    return this.getCachedJSON(`policy:${policyId}`);
  }

  async cacheSession(sessionId: string, data: Record<string, unknown>, ttl: number = 1800): Promise<void> {
    await this.cacheJSON(`session:${sessionId}`, data, ttl);
  }

  async getSession(sessionId: string): Promise<Record<string, unknown> | null> {
    return this.getCachedJSON(`session:${sessionId}`);
  }

  async close(): Promise<void> {
    if (this.client) await this.client.quit();
  }
}
